File: Server/src/main/java/com/github/dr/rwserver/command/ClientCommands.java
package com.github.dr.rwserver.command;

import com.github.dr.rwserver.command.ex.Vote;
import com.github.dr.rwserver.core.Call;
import com.github.dr.rwserver.core.thread.Threads;
import com.github.dr.rwserver.data.Player;
import com.github.dr.rwserver.data.global.Data;
import com.github.dr.rwserver.ga.GroupGame;
import com.github.dr.rwserver.game.EventType;
import com.github.dr.rwserver.game.GameMaps;
import com.github.dr.rwserver.game.Rules;
import com.github.dr.rwserver.game.Team;
import com.github.dr.rwserver.util.LocaleUtil;
import com.github.dr.rwserver.util.Time;
import com.github.dr.rwserver.util.game.CommandHandler;
import com.github.dr.rwserver.util.game.Events;
import com.github.dr.rwserver.util.log.Log;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import static com.github.dr.rwserver.data.global.Data.LINE_SEPARATOR;
import static com.github.dr.rwserver.game.GameMaps.MapType;
import static com.github.dr.rwserver.util.IsUtil.*;

/**
 * @author Dr
 */
public class ClientCommands {
	private final LocaleUtil localeUtil = Data.localeUtil;

	public ClientCommands(CommandHandler handler) {
		handler.<Player>register("help", "clientCommands.help", (args, player) -> {
			StringBuilder str = new StringBuilder(16);
			for(CommandHandler.Command command : handler.getCommandList()){
				if (command.description.startsWith("#")) {
					str.append("   ").append(command.text).append(command.paramText.isEmpty() ? "" : " ").append(command.paramText).append(" - ").append(command.description.substring(1));
				} else {
					if ("HIDE".equals(command.description)) {
						continue;
					}
					str.append("   ").append(command.text).append(command.paramText.isEmpty() ? "" : " ").append(command.paramText).append(" - ").append(player.localeUtil.getinput(command.description))
							.append(LINE_SEPARATOR);
				}
			}
			player.sendSystemMessage(str.toString());
		});

		handler.<Player>register("map", "<MapNumber...>", "clientCommands.map", (args, player) -> {
			if (isAdmin(player)) {
				if ( GroupGame.games.get(player.groupId).isStartGame ||  GroupGame.games.get(player.groupId).mapLock) {
					return;
				}
				final StringBuilder response = new StringBuilder(args[0]);
				for(int i=1,lens=args.length;i<lens;i++) {
					response.append(" ").append(args[i]);
				}
				final String inputMapName = response.toString().replace("'", "").replace(" ", "").replace("-", "").replace("_", "");
				final String mapPlayer = Data.MapsMap.get(inputMapName);
				if (mapPlayer != null) {
					String[] data = mapPlayer.split("@");
					 GroupGame.games.get(player.groupId).maps.mapName = data[0];
					 GroupGame.games.get(player.groupId).maps.mapPlayer = data[1];
					 GroupGame.games.get(player.groupId).maps.mapType = MapType.defaultMap;
				} else {
					if ( GroupGame.games.get(player.groupId).mapsData.size == 0) {
						return;
					}
					if (notIsNumeric(inputMapName)) {
						player.sendSystemMessage(localeUtil.getinput("err.noNumber"));
						return;
					}
					String name =  GroupGame.games.get(player.groupId).mapsData.keys().toSeq().get(Integer.parseInt(inputMapName));
					GameMaps.MapData data =  GroupGame.games.get(player.groupId).mapsData.get(name);
					 GroupGame.games.get(player.groupId).maps.mapData = data;
					 GroupGame.games.get(player.groupId).maps.mapType = data.mapType;
					 GroupGame.games.get(player.groupId).maps.mapName = name;
					 GroupGame.games.get(player.groupId).maps.mapPlayer = "";
					player.sendSystemMessage(player.localeUtil.getinput("map.custom.info"));
				}
				Call.upDataGameData();
			}
		});

		handler.<Player>register("maps", "[page]", "clientCommands.maps", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(localeUtil.getinput("err.startGame"));
				return;
			}
			if ( GroupGame.games.get(player.groupId).mapsData.size == 0) {
				return;
			}
			StringBuilder response = new StringBuilder();
			final AtomicInteger i = new AtomicInteger(0);
			response.append(localeUtil.getinput("maps.top"));
			 GroupGame.games.get(player.groupId).mapsData.each((k,v) -> {
				response.append(localeUtil.getinput("maps.info", i.get(),k)).append(LINE_SEPARATOR);
				i.getAndIncrement();
			});
			player.sendSystemMessage(response.toString());
		});

		handler.<Player>register("afk", "clientCommands.afk", (args, player) -> {
			if (! GroupGame.games.get(player.groupId).isAfk) {
				player.sendSystemMessage(localeUtil.getinput("ban.comm","afk"));
				return;
			}
			if (player.isAdmin) {
				player.sendSystemMessage(localeUtil.getinput("afk.adminNo"));
			} else {
				if ( GroupGame.games.get(player.groupId).isStartGame) {
					player.sendSystemMessage(localeUtil.getinput("err.startGame"));
					return;
				}
				if (Threads.getIfScheduledFutureData("AfkCountdown")) {
					return;
				}
				AtomicBoolean admin = new AtomicBoolean(true);
				Data.playerGroup.each(p -> p.isAdmin,e -> admin.set(false));
				if (admin.get()&& GroupGame.games.get(player.groupId).oneAdmin)  {
					player.isAdmin = true;
					Call.upDataGameData();
					Call.sendSystemMessageLocal("afk.end.noAdmin",player.groupId,player.name);
					return;
				}
				Threads.newThreadService(() -> Data.playerGroup.each(p -> p.isAdmin, i -> {
					i.isAdmin = false;
					player.isAdmin = true;
					Call.upDataGameData();
					Call.sendSystemMessageLocal("afk.end.ok",player.groupId,player.name);
					Threads.removeScheduledFutureData("AfkCountdown");
				}),30, TimeUnit.SECONDS,"AfkCountdown");
				Call.sendMessageLocal(player,"afk.start",player.name);
			}
		});

		handler.<Player>register("give", "<PlayerSerialNumber>", "clientCommands.give", (args, player) -> {
			if (isAdmin(player)) {
				if ( GroupGame.games.get(player.groupId).isStartGame) {
					player.sendSystemMessage(localeUtil.getinput("err.startGame"));
					return;
				}
				if (notIsNumeric(args[0])) {
					player.sendSystemMessage(localeUtil.getinput("err.noNumber"));
					return;
				}
				final int playerSite = Integer.parseInt(args[0])-1;
				final Player newAdmin =  GroupGame.games.get(player.groupId).playerData[playerSite];
				if (notIsBlank(newAdmin)) {
					player.isAdmin = false;
					newAdmin.isAdmin = true;
					Call.upDataGameData();
					Call.sendMessageLocal(player,"give.ok",player.name);
				}else{
					Call.sendMessageLocal(player,"give.noPlayer",player.name);
				}
			}
		});

		handler.<Player>register("nosay", "<on/off>", "clientCommands.noSay", (args, player) -> {
			player.noSay = "on".equals(args[0]);
			player.sendSystemMessage(localeUtil.getinput("server.noSay",(player.noSay) ? "开启" : "关闭"));
		});

		handler.<Player>register("am", "<on/off>", "clientCommands.am", (args, player) -> {
			 GroupGame.games.get(player.groupId).amTeam = "on".equals(args[0]);
			if ( GroupGame.games.get(player.groupId).amTeam) {
				Team.amYesPlayerTeam(player.groupId);
			} else {
				Team.amNoPlayerTeam(player.groupId);
			}
			player.sendSystemMessage(localeUtil.getinput("server.amTeam",( GroupGame.games.get(player.groupId).amTeam) ? "开启" : "关闭"));
		});

		handler.<Player>register("income", "<income>", "clientCommands.income", (args, player) -> {
			if (isAdmin(player)) {
				if ( GroupGame.games.get(player.groupId).isStartGame) {
					player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
					return;
				}
				float v = Float.parseFloat(args[0]);
				if(v<100) player.con.sendSystemMessage("已锁定不小于100倍金钱");
				else {
					GroupGame.games.get(player.groupId).income =v ;
					Call.upDataGameData();

				}
			}
		});

		handler.<Player>register("status", "clientCommands.status", (args, player) -> player.sendSystemMessage(player.localeUtil.getinput("status.version",Data.playerGroup.size(),Data.core.admin.bannedIPs.size(),Data.SERVER_CORE_VERSION,player.con.getVersion())));

		handler.<Player>register("kick", "<PlayerSerialNumber>", "clientCommands.kick", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
				return;
			}
			if (isAdmin(player)) {
				if (notIsNumeric(args[0])) {
					player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
					return;
				}
				final int site = Integer.parseInt(args[0])-1;
				if ( GroupGame.games.get(player.groupId).playerData[site] != null) {
					 GroupGame.games.get(player.groupId).playerData[site].kickTime = Time.getTimeFutureMillis(60 * 1000L);
					try {
						 GroupGame.games.get(player.groupId).playerData[site].con.sendKick(localeUtil.getinput("kick.you"));
					} catch (IOException e) {
						Log.error("[Player] Send Kick Player Error",e);
					}
				}
			}
		});

		handler.<Player>register("i", "<i...>","HIDE", (args, player) -> {
		});

		/* QC */
		handler.<Player>register("credits", "<money>", "HIDE", (args, player) -> {
			if (isAdmin(player)) {
				if (notIsNumeric(args[0])) {
					player.sendSystemMessage(localeUtil.getinput("err.noNumber"));
					return;
				}
				switch (Integer.parseInt(args[0])) {
					case 0:
						 GroupGame.games.get(player.groupId).credits = 1;
						break;
					case 1000:
						 GroupGame.games.get(player.groupId).credits = 2;
						break;
					case 2000:
						 GroupGame.games.get(player.groupId).credits = 3;
						break;
					case 5000:
						 GroupGame.games.get(player.groupId).credits = 4;
						break;
					case 10000:
						 GroupGame.games.get(player.groupId).credits = 5;
						break;
					case 50000:
						 GroupGame.games.get(player.groupId).credits = 6;
						break;
					case 100000:
						 GroupGame.games.get(player.groupId).credits = 7;
						break;
					case 200000:
						 GroupGame.games.get(player.groupId).credits = 8;
						break;
					case 4000:
						 GroupGame.games.get(player.groupId).credits = 0;
						break;
					default:
						break;

				}
				Call.upDataGameData();
			}
		});

		handler.<Player>register("nukes", "<boolean>", "HIDE", (args, player) -> {
			if (isAdmin(player)) {
				 GroupGame.games.get(player.groupId).noNukes = !Boolean.parseBoolean(args[0]);
				Call.upDataGameData();
			}
		});

		handler.<Player>register("addai", "HIDE", (args, player) -> player.sendSystemMessage(player.localeUtil.getinput("err.nosupr")));

		handler.<Player>register("fog", "<type>", "HIDE", (args, player) -> {
			if (isAdmin(player)) {
				 GroupGame.games.get(player.groupId).mist = "off".equals(args[0]) ? 0 : "basic".equals(args[0]) ? 1 : 2;
				Call.upDataGameData();
			}
		});

		handler.<Player>register("sharedcontrol", "<boolean>", "HIDE", (args, player) -> {
			if (isAdmin(player)) {
				 GroupGame.games.get(player.groupId).sharedControl = Boolean.parseBoolean(args[0]);
				Call.upDataGameData();
			}
		});

		handler.<Player>register("startingunits", "<type>", "HIDE", (args, player) -> {
			if (isAdmin(player)) {
				if (notIsNumeric(args[0])) {
					player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
					return;
				}
				// GroupGame.games.get(player.groupId).initUnit = (type == 1) ? 1 : (type == 2) ? 2 : (type ==3) ? 3 : (type == 4) ? 4 : 100;
				 GroupGame.games.get(player.groupId).initUnit = Integer.parseInt(args[0]);
				Call.upDataGameData();
			}
		});

		handler.<Player>register("start", "clientCommands.start", (args, player) -> {
			if (isAdmin(player)) {
				if (Threads.getIfScheduledFutureData("AfkCountdown")) {
					Threads.removeScheduledFutureData("AfkCountdown");
					Call.sendMessageLocal(player, "afk.clear", player.name);
				}
				Rules rules = GroupGame.games.get(player.groupId);
				if ( rules.startMinPlayerSize > GroupGame.playerGroup(player.groupId).size()) {
					player.sendSystemMessage(player.localeUtil.getinput("start.playerNo", rules.startMinPlayerSize));
					return;
				}
//				if (Threads.getIfScheduledFutureData("GamePing")) {
//					Threads.removeScheduledFutureData("GamePing");
//				}
				if ( rules.maps.mapData != null) {
					 rules.maps.mapData.readMap();
				}
				GroupGame.removePlayer(Data.playerAll,player.groupId);
				Data.playerGroup.eachBooleanIfs(p->p.groupId==player.groupId,e -> {
					try {
						e.con.sendStartGame();
						e.lastMoveTime = System.currentTimeMillis();
						Data.playerAll.add(e);
					} catch (IOException err) {
						Log.error("Start Error",err);
					}
				});

				if ( rules.winOrLose) {

				}
				rules.isStartGame = true;
				int int3 = 0;
				for (int i = 0; i <  rules.maxPlayer; i++) {
					Player player1 =  rules.playerData[i];
					if (player1 != null) {
						if (player1.sharedControl ||  rules.sharedControl) {
							int3 = (int3 | 1 << i);
						}
					}
				}
				rules.sharedControlPlayer = int3;
				Call.testPreparationPlayer(player.groupId);
				Events.fire(new EventType.GameStartEvent(player.groupId));
			}
		});

		handler.<Player>register("t", "<text...>","clientCommands.t", (args, player) -> {
			final StringBuilder response = new StringBuilder(args[0]);
			for(int i=1,lens=args.length;i<lens;i++) {
				response.append(" ").append(args[i]);
			}
			Call.sendTeamMessage(player.team,player,response.toString());
		});

		handler.<Player>register("surrender", "clientCommands.surrender", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				if (isBlank(Data.Vote)) {
					Data.Vote = new Vote(player,"surrender");
				} else {
					Data.Vote.toVote(player,"y");
				}
			} else {
				player.sendSystemMessage(player.localeUtil.getinput("err.noStartGame"));
			}
		});

		handler.<Player>register("teamlock", "clientCommands.teamlock", (args, player) -> {
			if (isAdmin(player)) {
				 GroupGame.games.get(player.groupId).lockTeam = "on".equals(args[0]);
			}
		});

		handler.<Player>register("killme", "clientCommands.killMe", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.con.sendSurrender();
			} else {
				player.sendSystemMessage(player.localeUtil.getinput("err.noStartGame"));
			}
		});

		handler.<Player>register("vote","<gameover/kick> [player-site]","clientCommands.vote", (args, player) -> {
			switch(args[0].toLowerCase()) {
				case "gameover":
					Data.Vote = new Vote(player, args[0]);
					break;
				case "kick":
					if (args.length > 1 && isNumeric(args[1])) {
						Data.Vote = new Vote(player, args[0], args[1]);
					} else {
						player.sendSystemMessage(player.localeUtil.getinput("err.commandError"));
					}
					break;
				default:
					player.sendSystemMessage(player.localeUtil.getinput("err.command"));
					break;
			}
		});

		handler.<Player>register("move", "<PlayerSerialNumber> <ToSerialNumber> <?>","HIDE", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
				return;
			}
			if (isAdmin(player)) {
				if (notIsNumeric(args[0]) && notIsNumeric(args[1]) && notIsNumeric(args[2])) {
					player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
					return;
				}
				int oldSite = Integer.parseInt(args[0])-1;
				int newSite = Integer.parseInt(args[1])-1;
				//int newSite = 0;
				int team = Integer.parseInt(args[2]);
				if (oldSite <  GroupGame.games.get(player.groupId).maxPlayer && newSite <  GroupGame.games.get(player.groupId).maxPlayer) {
					final Player od =  GroupGame.games.get(player.groupId).playerData[oldSite];
					if (newSite > -2) {
						if ( GroupGame.games.get(player.groupId).playerData[newSite] == null) {
							 GroupGame.games.get(player.groupId).playerData[oldSite] = null;
							od.site = newSite;
							if (team > -1) {
								od.team = team;
							}
							 GroupGame.games.get(player.groupId).playerData[newSite] = od;
						} else {
							final Player nw =  GroupGame.games.get(player.groupId).playerData[newSite];
							od.site = newSite;
							nw.site = oldSite;
							if (team >-1) {
								od.team = team;
							}
							 GroupGame.games.get(player.groupId).playerData[newSite] = od;
							 GroupGame.games.get(player.groupId).playerData[oldSite] = nw;
						}
					}
					Call.sendTeamData(player.groupId);
				}
			}
		});

		handler.<Player>register("self_move", "<ToSerialNumber> <?>","HIDE", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
				return;
			}
			if ( GroupGame.games.get(player.groupId).lockTeam) {
				return;
			}
			if (notIsNumeric(args[0]) && notIsNumeric(args[1])) {
				player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
				return;
			}
			int newSite = Integer.parseInt(args[0])-1;
			int team = Integer.parseInt(args[1]);
			if (newSite <  GroupGame.games.get(player.groupId).maxPlayer) {
				if ( GroupGame.games.get(player.groupId).playerData[newSite] == null) {
					 GroupGame.games.get(player.groupId).playerData[player.site] = null;
					player.site = newSite;
					if (team >-1) {
						player.team = team;
					}
					 GroupGame.games.get(player.groupId).playerData[newSite] = player;
					Call.sendTeamData(player.groupId);
				}
			}
		});

		handler.<Player>register("team", "<PlayerSiteNumber> <ToTeamNumber>","HIDE", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
				return;
			}
			if (isAdmin(player)) {
				if (notIsNumeric(args[0]) && notIsNumeric(args[1])) {
					player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
					return;
				}
				int playerSite = Integer.parseInt(args[0])-1;
				int newSite = Integer.parseInt(args[1])-1;
				if (playerSite <  GroupGame.games.get(player.groupId).maxPlayer && newSite <  GroupGame.games.get(player.groupId).maxPlayer) {
					if (newSite > -1) {
						if (notIsBlank( GroupGame.games.get(player.groupId).playerData[playerSite])) {
							 GroupGame.games.get(player.groupId).playerData[playerSite].team = newSite;
						}
					}
					Call.sendTeamData(player.groupId);
				}
			}
		});

		handler.<Player>register("self_team", "<ToTeamNumber>","HIDE", (args, player) -> {
			if ( GroupGame.games.get(player.groupId).isStartGame) {
				player.sendSystemMessage(player.localeUtil.getinput("err.startGame"));
				return;
			}
			if ( GroupGame.games.get(player.groupId).lockTeam) {
				return;
			}
			if (notIsNumeric(args[0])) {
				player.sendSystemMessage(player.localeUtil.getinput("err.noNumber"));
				return;
			}
			int newSite = Integer.parseInt(args[0])-1;
			if (newSite < GroupGame.games.get(player.groupId).maxPlayer) {
				player.team = newSite;
				Call.sendTeamData(player.groupId);
			}
		});


	}

	private boolean isAdmin(Player player) {
		if (player.isAdmin) {
			return true;
		}
		player.sendSystemMessage(player.localeUtil.getinput("err.noAdmin"));
		return false;
	}
}

File: Server-Core/src/main/java/net/rwhps/server/game/simulation/gameFramework/command/ClientCommands.kt
/*
 * Copyright 2020-2023 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package net.rwhps.server.game.simulation.gameFramework.command

import com.corrodinggames.rts.game.n
import com.corrodinggames.rts.gameFramework.j.ai
import net.rwhps.server.core.thread.CallTimeTask
import net.rwhps.server.core.thread.Threads
import net.rwhps.server.data.HessModuleManage
import net.rwhps.server.data.global.Data
import net.rwhps.server.data.global.Data.LINE_SEPARATOR
import net.rwhps.server.data.player.AbstractPlayer
import net.rwhps.server.data.plugin.PluginManage
import net.rwhps.server.game.simulation.gameFramework.GameEngine
import net.rwhps.server.util.IsUtil.notIsNumeric
import net.rwhps.server.util.game.CommandHandler
import net.rwhps.server.util.log.Log.error
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * @author RW-HPS/Dr
 */
internal class ClientCommands(handler: CommandHandler) {
    private val localeUtil = Data.i18NBundle

    private fun isAdmin(player: AbstractPlayer): Boolean {
        if (player.isAdmin) {
            return true
        }
        player.sendSystemMessage(player.i18NBundle.getinput("err.noAdmin"))
        return false
    }

    private fun checkSiteNumb(int: String, player: AbstractPlayer): Boolean {
        if (notIsNumeric(int)) {
            player.sendSystemMessage(player.i18NBundle.getinput("err.noNumber"))
            return false
        }
        if (int.toInt() < 0) {
            player.sendSystemMessage(player.i18NBundle.getinput("err.noInt"))
            return false
        }
        if (int.toInt() > (Data.config.MaxPlayer)) {
            player.sendSystemMessage(player.i18NBundle.getinput("err.maxPlayer"))
            return false
        }
        return true
    }

    init {
        val room = HessModuleManage.hessLoaderMap[this::class.java.classLoader.toString()].room
        
        handler.register("help", "clientCommands.help") { _: Array<String>?, player: AbstractPlayer ->
            val str = StringBuilder(16)
            for (command in handler.commandList) {
                if (command.description.startsWith("#")) {
                    str.append("   ").append(command.text).append(if (command.paramText.isEmpty()) "" else " ")
                        .append(command.paramText).append(" - ").append(command.description.substring(1))
                } else {
                    if ("HIDE" == command.description) {
                        continue
                    }
                    str.append("   ").append(command.text).append(if (command.paramText.isEmpty()) "" else " ")
                        .append(command.paramText).append(" - ").append(player.i18NBundle.getinput(command.description))
                        .append(LINE_SEPARATOR)
                }
            }
            player.sendSystemMessage(str.toString())
        }
        handler.register("map", "<MapNumber...>", "clientCommands.map") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                if (room.isStartGame) {
                    return@register
                }
                val response = StringBuilder(args[0])
                var i = 1
                val lens = args.size
                while (i < lens) {
                    response.append(" ").append(args[i])
                    i++
                }
                val inputMapName = response.toString().replace("'", "").replace(" ", "").replace("-", "").replace("_", "")
                val mapPlayer = Data.MapsMap[inputMapName]
                if (mapPlayer != null) {
                    val data = mapPlayer.split("@").toTypedArray()
                    GameEngine.netEngine.az = "maps/skirmish/${data[1]}${data[0]}.tmx"
                    GameEngine.netEngine.ay.b = "${data[1]}${data[0]}.tmx".also {
                        room.mapName = it
                    }
                    GameEngine.netEngine.ay.a = ai.a
                }
                room.call.sendSystemMessage(localeUtil.getinput("map.to",player.name,Data.game.maps.mapName))
                GameEngine.netEngine.L()
            }
        }
        handler.register("afk", "clientCommands.afk") { _: Array<String>?, player: AbstractPlayer ->
            if (!room.isAfk) {
                player.sendSystemMessage(player.i18NBundle.getinput("ban.comm", "afk"))
                return@register
            }
            if (player.isAdmin) {
                player.sendSystemMessage(player.i18NBundle.getinput("afk.adminNo"))
            } else {
                if (room.isStartGame) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                    return@register
                }
                if (Threads.containsTimeTask(CallTimeTask.PlayerAfkTask)) {
                    return@register
                }
                val admin = AtomicBoolean(true)
                room.playerManage.playerGroup.eachAllFind({ p: AbstractPlayer -> p.isAdmin },{ _: AbstractPlayer -> admin.set(false) })
                if (admin.get()) {
                    player.isAdmin = true
                    room.call.sendSystemMessageLocal("afk.end.noAdmin", player.name)
                    return@register
                }
                Threads.newCountdown(CallTimeTask.PlayerAfkTask, 30, TimeUnit.SECONDS) {
                    room.playerManage.playerGroup.eachFind(
                        { p: AbstractPlayer -> p.isAdmin }) { i: AbstractPlayer ->
                        i.isAdmin = false
                        player.isAdmin = true
                        room.call.sendSystemMessageLocal("afk.end.ok", player.name)
                        GameEngine.netEngine.L()
                    }
                }
                room.call.sendSystemMessageLocal("afk.start", player.name)
            }
        }
        handler.register("income", "<income>", "clientCommands.income") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                if (room.isStartGame) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.startGame.warn"))
                }
                GameEngine.netEngine.ay.h = args[0].toFloat()
            }
        }
        handler.register("addmoney", "<PlayerSerialNumber> <money>", "clientCommands.addmoney") { args: Array<String>, player: AbstractPlayer ->
            if (!room.isStartGame) {
                player.sendSystemMessage(player.i18NBundle.getinput("err.noStartGame"))
                return@register
            }
            if (isAdmin(player)) {
                if (notIsNumeric(args[0]) || notIsNumeric(args[1])) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.noNumber"))
                    return@register
                }

                val site = args[0].toInt() - 1
                val addMoneyPlayer = room.playerManage.getPlayerArray(site)
                if (addMoneyPlayer != null) {
                    try {
                        addMoneyPlayer.credits += args[1].toInt()
                    } catch (e: IOException) {
                        error("[Player] addMoneyPlayer Error", e)
                    }
                }
            }
        }
        handler.register("ai", "<difficuld>", "clientCommands.income") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                if (room.isStartGame) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                    return@register
                }
                GameEngine.netEngine.ay.f = args[0].toInt()
            }
        }
        handler.register("status", "clientCommands.status") { _: Array<String>?, player: AbstractPlayer ->
            player.sendSystemMessage(player.i18NBundle.getinput(
                "status.version",
                room.playerManage.playerGroup.size,
                0,
                Data.SERVER_CORE_VERSION,
                "RW-HPS-Hess"
            ))
        }
        handler.register("kick", "<PlayerSerialNumber>", "clientCommands.kick") { args: Array<String>, player: AbstractPlayer ->
            if (room.isStartGame) {
                player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                return@register
            }
            if (isAdmin(player)) {
                if (notIsNumeric(args[0])) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.noNumber"))
                    return@register
                }

                val site = args[0].toInt() - 1
                val kickPlayer = room.playerManage.getPlayerArray(site)
                if (kickPlayer != null) {
                    try {
                        kickPlayer.kickPlayer(localeUtil.getinput("kick.you"),60)
                    } catch (e: IOException) {
                        error("[Player] Send Kick Player Error", e)
                    }
                }
            }
        }

        /* QC */
        handler.register("credits", "<money>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                if (notIsNumeric(args[0])) {
                    player.sendSystemMessage(localeUtil.getinput("err.noNumber"))
                    return@register
                }
                when (args[0].toInt()) {
                    0 -> GameEngine.netEngine.ay.c = 1
                    1000 -> GameEngine.netEngine.ay.c = 2
                    2000 -> GameEngine.netEngine.ay.c = 3
                    5000 -> GameEngine.netEngine.ay.c = 4
                    10000 -> GameEngine.netEngine.ay.c = 5
                    50000 -> GameEngine.netEngine.ay.c = 6
                    100000 -> GameEngine.netEngine.ay.c = 7
                    200000 -> GameEngine.netEngine.ay.c = 8
                    4000 -> GameEngine.netEngine.ay.c = 0
                    else -> {
                    }
                }
            }
        }
        handler.register("nukes", "<boolean>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                GameEngine.netEngine.ay.i = !args[0].toBoolean()
            }
        }
        handler.register("addai", "HIDE") { _: Array<String>?, player: AbstractPlayer ->
            if (isAdmin(player)) {
                GameEngine.root.multiplayer.addAI()
            }
        }
        handler.register("fog", "<type>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                GameEngine.netEngine.ay.d = if ("off" == args[0]) 0 else if ("basic" == args[0]) 1 else 2
            }
        }
        handler.register("sharedcontrol", "<boolean>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                GameEngine.netEngine.ay.l = args[0].toBoolean()
            }
        }
        handler.register("startingunits", "<type>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (isAdmin(player)) {
                GameEngine.netEngine.ay.g = args[0].toInt()
            }
        }
        handler.register("start", "clientCommands.start") { _: Array<String>?, player: AbstractPlayer ->
            if (isAdmin(player)) {
                if (room.isStartGame) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                    return@register
                }

                GameEngine.root.multiplayer.multiplayerStart()
            }
        }
        handler.register("move", "<PlayerSerialNumber> <ToSerialNumber> <Team>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (room.isStartGame) {
                player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                return@register
            }
            if (isAdmin(player)) {
                if (notIsNumeric(args[0]) && notIsNumeric(args[1]) && notIsNumeric(args[2])) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.noNumber"))
                    return@register
                }

                val tg = args[0].toInt() - 1
                if (tg == 10) {
                    player.sendSystemMessage("禁止操作此玩家")
                    return@register
                }
                val player = n.k(tg)
                val site = args[1].toInt() - 1
                val newTeam = args[2].toInt()
                GameEngine.netEngine.a(player, site)
                if (newTeam == -1) {
                    player.r = site % 2;
                } else {
                    player.r = newTeam
                }
            }
        }
        handler.register("team", "<PlayerSiteNumber> <ToTeamNumber>", "HIDE") { args: Array<String>, player: AbstractPlayer ->
            if (room.isStartGame) {
                player.sendSystemMessage(player.i18NBundle.getinput("err.startGame"))
                return@register
            }
            if (isAdmin(player)) {
                if (notIsNumeric(args[0]) && notIsNumeric(args[1])) {
                    player.sendSystemMessage(player.i18NBundle.getinput("err.noNumber"))
                    return@register
                }
                val playerSite = args[0].toInt() - 1
                val newSite = args[1].toInt() - 1
                if (playerSite == 10) {
                    player.sendSystemMessage("禁止操作此玩家")
                    return@register
                }
                n.k(playerSite).r = newSite
            }
        }

        PluginManage.runRegisterServerClientCommands(handler)
    }
}